### ISSUE TITLE
SSLSocketInitiator does not close socket if handshakeSSL fails

### ISSUE THREAD

**Issue by leleftheriades** [quickfix/quickfix#284]
In method `SSLSocketInitiator::onConnect` when `sslHandshakeSuccess == false`
`pSocketConnection` is deleted, but socket may not be closed and more importantly it is not dropped from `SocketMonitor`.
This results in `SocketMonitor::block`'s call to `select` to fail because an invalid socket handle is left in the read set.

A solution would be to call pSocketConnection->disconnect(); before deleting the pSocketConnection object in the else branch of `SSLSocketInitiator::onConnect`

**Comment by leleftheriades** [comment]
Correction: to fix this issue Session::disconnect(); must be called. 
This will also let the session know that the connection is no longer valid so it does not attempt to send a logout message on the old connection that we will delete, when we assign the new connection.

**Comment by leleftheriades** [comment]
Possible side-effects from this bug include:
-Permanent failure to connect to desired target because the monitor select will always fail
-Application crash because session will try to access the previous connection that was deleted

**Comment by leleftheriades** [comment]
Suggested fix:
```
void SSLSocketInitiator::onConnect( SocketConnector& connector, socket_handle s )
{
  getLog()->onEvent("Socket connected handle: " + std::to_string(s));

  SocketConnections::iterator i = m_pendingConnections.find( s );
  if( i == m_pendingConnections.end() ) return;
  SSLSocketConnection* pSocketConnection = i->second;

  bool sslHandshakeSuccess = handshakeSSL(pSocketConnection->sslObject());

  if (sslHandshakeSuccess) {
      m_connections[s] = pSocketConnection;
      m_pendingConnections.erase(i);
      setConnected(pSocketConnection->getSession()->getSessionID());
      pSocketConnection->onTimeout();
  }else
  {
      setDisconnected(pSocketConnection->getSession()->getSessionID());      
	  
	  Session* pSession = pSocketConnection->getSession();
	  if (pSession)
	  {
		  pSession->disconnect();
		  setDisconnected(pSession->getSessionID());
	  }

      delete pSocketConnection;
	  m_pendingConnections.erase(i);

      getLog()->onEvent("Socket deleted due to ssl handshake error");
  }
}
```

**Comment by leleftheriades** [comment]
I have included this fix in the pull request #274 which fixes another ssl issue.
